README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cfdecomp v0.4.0

<!-- badges: start -->

[![Build
Status](https://travis-ci.com/MaartenBijlsma/cfdecomp.svg?branch=master)](https://travis-ci.com/MaartenBijlsma/cfdecomp)
<!-- badges: end -->

cfdecomp stands for ‘counterfactual decomposition’. This package
provides functions to decompose differences in an outcome attributable
to a mediating variable (or sets of mediating variables) between groups
based on counterfactual (causal inference) theory. By using Monte Carlo
(MC) integration (simulations based on empirical estimates from
multivariable models) we provide added flexibility compared to existing
(analytical) approaches, at the cost of computational power or time. The
added flexibility means that we can decompose difference between groups
in any outcome or and with any mediator (any variable type and
distribution). See [Sudharsanan & Bijlsma
(2019)](https://dx.doi.org/10.4054/MPIDR-WP-2019-004) for more
information.

## Installation

You can install the released version of cfdecomp from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("cfdecomp")
```

## Example

``` r
library(cfdecomp)
#> Warning: package 'cfdecomp' was built under R version 4.0.3

formula.y.example <- 'out.gauss ~ med.gauss + age + med.binom'
formula.m.example <- 'med.gauss ~ age'

# run the procedure for a gaussian outcome and gaussian mediator
mean.results.1 <- cfd.mean(formula.y='out.gauss ~ SES + med.gauss + med.binom + age',
              formula.m='med.gauss ~ SES + age',
              mediator='med.gauss',
              group='SES',
              data=cfd.example.data,
              family.y='gaussian',
              family.m='gaussian',
              bs.size=250,
              mc.size=50,
              alpha=0.05)

# see if our modelling procedure approximates the mean of the mediator well
# for SES group 1
hist(mean.results.1$out_nc_m[,1],col='grey',xlab='MC iteration means of M of SES group 1',
     main='Comparison of MC approximation to empirical mean of mediator M')
# place empirical mean of M in the histogram:
abline(v=mean(cfd.example.data$med.gauss[cfd.example.data$SES==1]),lwd=4)
```

<img src="man/figures/README-example-1.png" width="100%" />

``` r
# approximation of the mean of M over the Monte Carlo iterations for SES group 1:
mean(mean.results.1$out_nc_m[,1])
#> [1] 8.160594
# empirical mean of mediator M for SES group 1:
mean(cfd.example.data$med.gauss[cfd.example.data$SES==1])
#> [1] 8.163461
# conclusion: on average, our distributions of means of M centers on the empirical distribution of M

# for SES group 2
hist(mean.results.1$out_nc_m[,2],col='grey',xlab='MC iteration means of M of SES group 2',
     main='Comparison of MC approximation to empirical mean of mediator M')
# place empirical mean of M in the histogram:
abline(v=mean(cfd.example.data$med.gauss[cfd.example.data$SES==2]),lwd=4)
```

<img src="man/figures/README-example-2.png" width="100%" />

``` r
mean(mean.results.1$out_nc_m[,2]);mean(cfd.example.data$med.gauss[cfd.example.data$SES==2])
#> [1] 7.179294
#> [1] 7.186521
# conclusion: also for group 2, approximation is good

# for SES group 3
hist(mean.results.1$out_nc_m[,3],col='grey',xlab='MC iteration means of M of SES group 3',
     main='Comparison of MC approximation to empirical mean of mediator M')
# place empirical mean of M in the histogram:
abline(v=mean(cfd.example.data$med.gauss[cfd.example.data$SES==3]),lwd=4)
```

<img src="man/figures/README-example-3.png" width="100%" />

``` r
mean(mean.results.1$out_nc_m[,3]);mean(cfd.example.data$med.gauss[cfd.example.data$SES==3])
#> [1] 5.377426
#> [1] 5.376548
# conclusion: also for group 3, approximation of M is good

# see if our modelling procedure approximates the mean of outcome Y well
# for SES group 1
hist(mean.results.1$out_nc_y[,1],col='grey',xlab='MC iteration means of Y of SES group 1',
     main='Comparison of MC approximation to empirical mean of outcome Y')
# place empirical mean of outcome Y in the histogram:
abline(v=mean(cfd.example.data$out.gauss[cfd.example.data$SES==1]),lwd=4)
```

<img src="man/figures/README-example-4.png" width="100%" />

``` r
mean(mean.results.1$out_nc_y[,1]);mean(cfd.example.data$out.gauss[cfd.example.data$SES==1])
#> [1] 4.143651
#> [1] 4.145881
# conclusion: on average, our distributions of means of Y centers on the empirical distribution of Y 

# for SES group 2
hist(mean.results.1$out_nc_y[,2],col='grey',xlab='MC iteration means of Y of SES group 2',
     main='Comparison of MC approximation to empirical mean of outcome Y')
# place empirical mean of outcome Y in the histogram:
abline(v=mean(cfd.example.data$out.gauss[cfd.example.data$SES==2]),lwd=4)
```

<img src="man/figures/README-example-5.png" width="100%" />

``` r
mean(mean.results.1$out_nc_y[,2]);mean(cfd.example.data$out.gauss[cfd.example.data$SES==2])
#> [1] 3.282936
#> [1] 3.284409
# conclusion: also for group 2, approximation of Y is good

# for SES group 3
hist(mean.results.1$out_nc_y[,3],col='grey',xlab='MC iteration means of Y of SES group 3',
     main='Comparison of MC approximation to empirical mean of outcome Y')
# place empirical mean of outcome Y in the histogram:
abline(v=mean(cfd.example.data$out.gauss[cfd.example.data$SES==3]),lwd=4)
```

<img src="man/figures/README-example-6.png" width="100%" />

``` r
mean(mean.results.1$out_nc_y[,3]);mean(cfd.example.data$out.gauss[cfd.example.data$SES==3])
#> [1] 2.224944
#> [1] 2.227222
# conclusion: also for group 3, approximation of Y is good

# looks good!

# estimate the effect of the intervention and proportion mediated (decomposition)
# the differences between SES groups 1 and 2 were first:
mean(mean.results.1$out_nc_y[,2] - mean.results.1$out_nc_y[,1])
#> [1] -0.8607148
# and after giving the gaussian mediator of SES group 2 the distribution of the one in group 1
# the difference becomes:
mean(mean.results.1$out_cf_y[,2] - mean.results.1$out_nc_y[,1])
#> [1] -0.6525706
# so the proportion of the outcome Y that is due to differences between the two SES groups in the gaussian mediator is
mean(1-(mean.results.1$out_cf_y[,2] - mean.results.1$out_nc_y[,1]) / (mean.results.1$out_nc_y[,2] - mean.results.1$out_nc_y[,1]))
#> [1] 0.2423695
# we can also get this number, and the one from the comparison of the other SES group with group 1, straight from the object
mean.results.1$mediation
#>         2         3 
#> 0.2423695 0.3060078
# you'll notice the first number is the same as the one we calculated ourselves from the output
# this is the proportion mediated when coming the first and the second SES group
# the second number is the proportion mediated when comparing the first and the third SES group
# and we can get the 1-alpha confidence intervals for each:
mean.results.1$mediation_quantile
#>               2         3
#> 2.5%  0.2073320 0.2793016
#> 97.5% 0.2869604 0.3320278

# if a mediator is binomial distributed
mean.results.2 <- cfd.mean(formula.y='out.gauss ~ med.pois + age + med.binom',
              formula.m='med.binom ~ age',
              mediator='med.binom',
              group='SES',
              data=cfd.example.data,
              family.y='gaussian',
              family.m='binomial',
              bs.size=250,
              mc.size=50,
              alpha=0.05)

# if a mediator is Poisson distributed
mean.results.3 <- cfd.mean(formula.y='out.gauss ~ med.pois + age + med.binom',
              formula.m='med.pois ~ age',
              mediator='med.pois',
              group='SES',
              data=cfd.example.data,
              family.y='gaussian',
              family.m='poisson',
              bs.size=250,
              mc.size=50,
              alpha=0.05)

# we have chosen 50 Monte Carlo iterations and 250 bootstrap iterations
# are these number enough for stability? (also consider the scale of the Y-axis)

# bootstrap, mediator, by SES group
plot(conv.mean(mean.results.1$out_nc_m[,1]),type='l',ylab='Cumulative average of M',
     main='Bootstrap stability in SES group 1',xlab='Bootstrap iteration')
```

<img src="man/figures/README-example-7.png" width="100%" />

``` r
plot(conv.mean(mean.results.1$out_nc_m[,2]),type='l',ylab='Cumulative average of M',
     main='Bootstrap stability in SES group 2',xlab='Bootstrap iteration')
```

<img src="man/figures/README-example-8.png" width="100%" />

``` r
plot(conv.mean(mean.results.1$out_nc_m[,3]),type='l',ylab='Cumulative average of M',
     main='Bootstrap stability in SES group 3',xlab='Bootstrap iteration')
```

<img src="man/figures/README-example-9.png" width="100%" />

``` r

# bootstrap, outcome, by SES group
plot(conv.mean(mean.results.1$out_nc_y[,1]),type='l',ylab='Cumulative average of Y',
     main='Bootstrap stability in SES group 1',xlab='Bootstrap iteration')
```

<img src="man/figures/README-example-10.png" width="100%" />

``` r
plot(conv.mean(mean.results.1$out_nc_y[,2]),type='l',ylab='Cumulative average of Y',
     main='Bootstrap stability in SES group 2',xlab='Bootstrap iteration')
```

<img src="man/figures/README-example-11.png" width="100%" />

``` r
plot(conv.mean(mean.results.1$out_nc_y[,3]),type='l',ylab='Cumulative average of Y',
     main='Bootstrap stability in SES group 3',xlab='Bootstrap iteration')
```

<img src="man/figures/README-example-12.png" width="100%" />

``` r
# a bootstrap size of 250 is more than enough for stability of the estimate
# of course, we want a large size to also get precise quantiles (confidence intervals)

# for mc iterations, let's first get some insight by going a higher mc.size number and only 1 bootstrap iteration
x <- cfd.mean(formula.y='out.gauss ~ med.pois + age + med.binom',
              formula.m='med.pois ~ age',
              mediator='med.pois',
              group='SES',
              data=cfd.example.data,
              family.y='gaussian',
              family.m='poisson',
              bs.size=1,
              mc.size=250,
              alpha=0.05)

# Monte Carlo, mediator, by SES group
plot(x$mc_conv_info_m[,1],type='l',ylab='Cumulative average of M',
     main='Monte Carlo stability in SES group 1',xlab='Monte Carlo iteration')
```

<img src="man/figures/README-example-13.png" width="100%" />

``` r
plot(x$mc_conv_info_m[,2],type='l',ylab='Cumulative average of M',
     main='Monte Carlo stability in SES group 2',xlab='Monte Carlo iteration')
```

<img src="man/figures/README-example-14.png" width="100%" />

``` r
plot(x$mc_conv_info_m[,3],type='l',ylab='Cumulative average of M',
     main='Monte Carlo stability in SES group 3',xlab='Monte Carlo iteration')
```

<img src="man/figures/README-example-15.png" width="100%" />

``` r

# Bootstrap, mediator, by SES group
plot(x$mc_conv_info_y[,1],type='l',ylab='Cumulative average of Y',
     main='Monte Carlo stability in SES group 1',xlab='Monte Carlo iteration')
```

<img src="man/figures/README-example-16.png" width="100%" />

``` r
plot(x$mc_conv_info_y[,2],type='l',ylab='Cumulative average of Y',
     main='Monte Carlo stability in SES group 2',xlab='Monte Carlo iteration')
```

<img src="man/figures/README-example-17.png" width="100%" />

``` r
plot(x$mc_conv_info_y[,3],type='l',ylab='Cumulative average of Y',
     main='Monte Carlo stability in SES group 3',xlab='Monte Carlo iteration')
```

<img src="man/figures/README-example-18.png" width="100%" />

``` r

# 50 Monte Carlo iterations seems very sufficient


# modelling some quantile of outcome Y (probs=0.50 = median)
quantile.results.1 <- cfd.quantile(formula.y='out.gauss ~ SES + med.gauss + med.binom + age',
                           formula.m='med.gauss ~ SES + age',
                           mediator='med.gauss',
                           group='SES',
                           data=cfd.example.data,
                           family.y='gaussian',
                           family.m='gaussian',
                           bs.size=250,
                           mc.size=50,
                           alpha=0.05,
                           probs=0.50)

# the distribution of the residuals of the mediator or outcome is very non-normal
# (in the example they are normal) we can also choose to sample from the residuals
# of their models rather than assume normality
quantile.results.2 <- cfd.quantile(formula.y='out.gauss ~ SES + med.gauss + med.binom + age',
                                   formula.m='med.gauss ~ SES + age',
                                   mediator='med.gauss',
                                   group='SES',
                                   data=cfd.example.data,
                                   family.y='gaussian',
                                   family.m='gaussian',
                                   bs.size=250,
                                   mc.size=50,
                                   alpha=0.05,
                                   probs=0.50,
                                   sample.resid.y=TRUE,
                                   sample.resid.m=TRUE)


# semi-parametric estimation ...
mean.semipar.results.1 <- cfd.semipar.mean(formula='out.gauss ~ SES + med.gauss + med.binom + age',
                           mediator='med.gauss',
                           group='SES',
                           strata='age',
                           nbin=5,
                           data=cfd.example.data,
                           family='gaussian',
                           bs.size=250,
                           mc.size=50,
                           alpha=0.05)

quantile.semipar.results.1 <- cfd.semipar.quantile(formula='out.gauss ~ SES + med.gauss + med.binom + age',
                                           mediator='med.gauss',
                                           group='SES',
                                           strata='age',
                                           nbin=5,
                                           data=cfd.example.data,
                                           family='gaussian',
                                           bs.size=250,
                                           mc.size=50,
                                           alpha=0.05,
                                           probs=0.50)

# for examples of the functions cluster.resample and cluster.id being used in cfd.mean and cfd.quantile
# please see the help files of those individual functions
# for an example of cfd.FUN, please see the help file of that individual function.
```


NEWS.md
# cfdecomp v0.4.0

This is the fourth release of the cfdecomp package. Compared to version 3, the cluster.resample function now makes use of rep() instead of a for() loop, speeding up the function considerably!


# cfdecomp v0.3.0

This is the third release of the cfdecomp package. Compared to version 2, this new version has a bug fix which only showed up when cluster resampling was used and due to sample size, no cluster was resampled more than once.

# cfdecomp v0.2.0

This is a second release of the cfdecomp package. Compared to version 1, this new version now has three added functions and added functionality to existing functions.

The added functions are:
* cluster.resample: a function that allows for easy resampling by cluster ID in long-format data.
* cfd.FUN: a decomposition function that can decompose the output of any function (as long as that function returns a vector)

The added functionality is:
* cluster.sample and cluster.name have been added to cfd.mean and cfd.quantile, allowing for cluster resampling in these functions.

See DESCRIPTION for more information.


R/cfd.example.data.R
#' @title Example Data for the cfdecomp package
#'
#' @description A dataframe with artificially generated data, intended to be used in a demonstration of the functions in this package.
#'
#' @format a dataframe with 6 columns, which are:
#' \describe{
#' \item{SES}{a factor variable with 3 levels, denoting three groups to be compared. SES stands for socio-economic status.}
#' \item{age}{a continuous variable going from 40 to 65. Age could stand for age in years.}
#' \item{med.gauss}{a mediator that is a continuous variable and normally distributed.}
#' \item{med.binom}{a mediator that is binomial.}
#' \item{med.pois}{a mediator that is a count variable and poisson distributed.}
#' \item{out.gauss}{an outcome that is a continuous variable and normally distributed.}
#' \item{id}{a variable with cluster id information}
#' }
"cfd.example.data"


R/cluster.resample.R

#' Cluster Resampling: resampling long format longitudinal or otherwise clustered data
#'
#' @param data a data frame containing the variables in the model.
#' @param cluster.name the name (as a character) of the column containing the cluster identifiers.
#' @param size a non-negative integer giving the number of items to choose, i.e. the number of clusters to resample. If not specified, takes the original data cluster size.
#' @param newID if set to TRUE, gives each replicant a new ID, rather than having the old ID appear multiple times. This might be needed if you work with e.g. cluster fixed effects
#'
#' @return returns a new dataframe with resampled clusters
#' @export
#'
#' @examples
#' table(cfd.example.data$id)
#' # every ID appears 5 times
#' cfd.example.sample <- cluster.resample(cfd.example.data, cluster.name='id')
#' table(cfd.example.sample$id)
#' # some ID's now don't appear, and some appear more times (multiples of 5)
#' # the important part is that if a person (id) is resampled, all their rows of data are taken
#' # i.e. the function resamples clusters, rather than rows.
#' # this has produced 1 resample, so generally this function would be used
#' # inside a loop where it is used multiple times
#' # we use this function inside our decomposition functions when cluster.sample is
#' # set to TRUE in those functions.
#'
cluster.resample <- function(data,cluster.name,size=NA,newID=FALSE) {

  # 1. sample cluster ids with replacement
  IDs <- unique(data[cluster.name])[,1]
  # if size has not been specified, use original cluster size
  if(is.na(size)) {size <- length(IDs)}
  y <- table(sample(IDs,size,replace=T))

  # 2. determine how many times a cluster id appears, save cluster id and how often it appeared
  id.rep <- data.frame(t(y))[,-1]
  names(id.rep) <- c(cluster.name,'freq')

  # 3. get all the observations for each cluster id in id.rep - # then right join id.rep onto data
  data.select <- data[which(data[cluster.name][,1] %in% id.rep[,1]),]
  data.merge <- merge(data.select,id.rep,all.x=TRUE)

  # remove unnecessary data from memory
  rm(data.select)

  # 4. use lapply to copy the rows
  data.sample <- as.data.frame(lapply(data.merge, rep, data.merge$freq))

  if(newID==TRUE) {
    # 5. label the replicates
    data.sample$replicate <- unlist(lapply(data.merge$freq,function(x) if(x > 0) 1:x))

    # 6. each clusterid & replication combination is the new unique id
    data.sample$newID <- paste(data.sample[cluster.name][,1],data.sample$replicate,sep='')

    # 7. order by newid
    data.sample[order(data.sample$newID),]

    # remove unneccesary data from memory
    data.sample$freq <- NULL
    data.sample$replicate <- NULL
  }
  rm(data.merge)

  # output the result
  return(data.sample)
}


R/cfd.FUN.R

#' Flexible Function Decomposition: decompose any function that returns a vector
#'
#' @param formula.y the \code{\link{formula}} for the multivariable model (see \code{\link{glm}}) for the outcome Y.
#' @param formula.m the \code{\link{formula}} for the multivariable model (see \code{\link{glm}}) for the mediator M.
#' @param mediator the column name of the mediator M.
#' @param group column name of a factor variable containing the group identifier.
#' @param data a data frame containing the variables in the model.
#' @param family.y a description of the error distribution to be used in the model, see \code{\link{family}} for details. For the outcome variable any member of the \code{glm} family can be used.
#' @param family.m a description of the error distribution to be used in the model, see \code{\link{family}} for details. For the mediator, currently \code{gaussian}, \code{binomial} and \code{poisson} are supported.
#' @param bs.size the number of bootstrap iterations to be performed.
#' @param mc.size the number of Monte Carlo iterations to be performed (more = more MC error reduction).
#' @param FUN.y a function to compute the statistics which can be applied to all data subsets, this function should return a vector and should be ran on pred_y (simulated y values in the natural course or counterfactual) and optional additional columns.
#' @param alpha the alpha level used to construct confidence intervals (0.05 = 95 percent confidence interval).
#' @param cluster.sample set to TRUE if data are clustered in the long format (i.e. multiple rows per individual or other cluster).
#' @param cluster.name the name (as a character) of the column containing the cluster identifiers.
#' @param cluster.mrows for the mediator model, only allows 1 observation per mediator so that the mediator model is not weighted by number of observations. e.g. set to TRUE if the mediator is time constant in longitudinal analysis of long format data.
#' @param sample.resid.y sample.resid if the \code{outcome} is Gaussian, should the simulation sample from the residuals of the linear regression model of the outcome to approximate the empirical distribution of the outcome in the simulation (Monte Carlo integration) (if so, set to \code{TRUE}), or should it sample from a Gaussian distribution with the standard deviation of the outcome? If the true distribution of the continuous outcome is not very Gaussian, the former may be preferred.
#' @param sample.resid.m sample.resid if the \code{mediator} is Gaussian, should the simulation sample from the residuals of the linear regression model of the mediator to approximate the empirical distribution of the mediator in the simulation (Monte Carlo integration) (if so, set to \code{TRUE}), or should it sample from a Gaussian distribution with the standard deviation of the mediator? If the true distribution of the continuous mediator is not very Gaussian, the former may be preferred.
#' @param print.iteration print the bootstrap iteration
#' @param ... further arguments passed to or used by methods.
#'
#' @return \code{out_nc_m} returns the mean level of the mediator under the natural course, which is a value that should be close to the empirically observed value of the mediator for each group. \code{out_nc_quantile} provides the \code{alpha/2} and \code{1-alpha/2} bootstrap quantiles for this mean (AKA bootstrap percentile confidence intervals). \code{out_nc_y} provides the output of the function fed into FUN.y for each bootstrap iteration, with \code{out_nc_quantile_y} providing the \code{alpha/2} and \code{1-alpha/2} bootstrap quantiles of that output. Similarly, \code{out_cf_m}, \code{out_cf_quantile_m},\code{out_cf_y}, and \code{out_cf_quantile_y} provide the corresponding values for the counterfactual scenario where the mediators of the groups are equalized. \code{mediation} and \code{mediation_quantile} are not provided for this function, so should be calculated by the user based on the output. \code{mc_conv_info_m} and \code{mc_conv_info_y} provide information that can help determine the number of Monte Carlo and Bootstrap iterations needed to achieve stability. See the \code{Examples} for more information.
#' @export
#'
#' @examples
#' set.seed(100)
#' # the decomposition functions in our package are computationally intensive
#' # to make the example run quick, I perform it on a subsample (n=250) of the data:
#' cfd.example.sample <- cfd.example.data[sample(250),]
#' # define some function (here one that calculates the mean from the data)
#' # such a function already exists, but this is to demonstrate how to do it for one that
#' # will be implemented in cfd.FUN:
#' mean.fun <- function(data,yname) {
#' x <- data
#' return(mean(x[,yname],na.rm=TRUE))
#' }
#' # test if the function works on normal data:
#' mean.fun(cfd.example.sample,yname="med.pois")
#' # then enter it into cfd.FUN and run:
#' mean.results <- cfd.FUN(formula.y='out.gauss ~ SES + med.gauss + med.binom + age',
#'                           formula.m='med.gauss ~ SES + age',
#'                           mediator='med.gauss',
#'                           group='SES',
#'                           data=cfd.example.sample,
#'                           family.y='gaussian',
#'                           family.m='gaussian',
#'                           FUN.y=mean.fun,
#'                           bs.size=15,
#'                           mc.size=5,
#'                           alpha=0.05,
#'                           print.iteration=TRUE,
#'                           yname="pred_y")
#' # more advanced code demonstrating how to do this with a function that calculates
#' # the age-adjusted rate ratio and life expectancy will hopefully soon be available
#' # in a publication.
#' #' @import stats utils
cfd.FUN <- function (formula.y, formula.m, mediator, group,data,
                     family.y = "binomial",
                     family.m = "binomial",
                     bs.size = 250,
                     mc.size = 50,
                     FUN.y=mean,
                     alpha = 0.05,
                     cluster.sample = FALSE,
                     cluster.name = NA,
                     cluster.mrows=FALSE,
                     sample.resid.y = FALSE,
                     sample.resid.m = FALSE,
                     print.iteration = FALSE,...)
{
  call <- match.call()
  if (missing(data))
    data <- environment(formula.y)
  mf <- match.call(expand.dots = FALSE)
  ini_fit.y <- glm(formula = formula.y, data = data, family = family.y)
  ini_fit.y$model_matrix <- model.matrix(as.formula(formula.y),
                                         data = data)
  fam.y <- family(ini_fit.y)
  if(cluster.mrows==FALSE) {
    ini_fit.m <- glm(formula = formula.m, data = data, family = family.m)
    ini_fit.m$model_matrix <- model.matrix(as.formula(formula.m),
                                           data = data)
  } else if(cluster.mrows==TRUE) {
    data.m <- do.call(rbind,by(data, list(data[,cluster.name]),FUN=function(x) head(x,1)))
    ini_fit.m <- glm(formula = formula.m, data = data.m, family = family.m)
    ini_fit.m$model_matrix <- model.matrix(as.formula(formula.m),
                                           data = data.m)
  }
  fam.m <- family(ini_fit.m)

  out_nc_m <- matrix(NA, nrow = bs.size, ncol = nlevels(data[,
                                                             group]), dimnames = list(1:bs.size, levels(data[, group])))
  out_cf_m <- matrix(NA, nrow = bs.size, ncol = nlevels(data[,
                                                             group]), dimnames = list(1:bs.size, levels(data[, group])))
  temp_nc_m <- matrix(NA, nrow = mc.size, ncol = nlevels(data[,
                                                              group]))
  temp_cf_m <- matrix(NA, nrow = mc.size, ncol = nlevels(data[,
                                                              group]))

  # ! let's determine the length of the matrix needed to store results from FUN.y
  test.data <- data

  pred_mean_y <- predict(ini_fit.y, test.data, type = "response")
  n <- length(pred_mean_y)
  if (family.y[1] == "gaussian" & sample.resid.y ==
      FALSE) {
    sd.ref.y <- sd(ini_fit.y$residuals[test.data[, group] ==
                                         levels(test.data[, group])[1]])
  }
  if (family.y[1] == "gaussian" & sample.resid.y ==
      TRUE) {
    resid.ref.y <- ini_fit.y$residuals[test.data[, group] ==
                                         levels(test.data[, group])[1]]
  }
  if (family.y[1] == "poisson") {
    pred_y <- rpois(n, pred_mean_y)
  }      else if (family.y[1] == "binomial") {
    pred_y <- rbinom(n, 1, pred_mean_y)
  }      else if (family.y[1] == "gaussian" & sample.resid.y ==
                  FALSE) {
    pred_y <- rnorm(n, pred_mean_y, sd = sd.ref.y)
  }      else if (family.y[1] == "gaussian" & sample.resid.y ==
                  TRUE) {
    pred_y <- pred_mean_y + sample(resid.ref.y, n,
                                   replace = TRUE)
  }
  test.data$pred_y <- pred_y
  test.vec <- FUN.y(test.data,...)
  rm(test.data)
  rm(pred_mean_y)
  vec.length <- length(test.vec)

  out_nc_y <- array(NA,dim=c(bs.size,vec.length))
  out_cf_y <- array(NA,dim=c(bs.size,vec.length))
  temp_nc_y <- array(NA,dim=c(mc.size,vec.length))
  temp_cf_y <- array(NA,dim=c(mc.size,vec.length))

  for (i in 1:bs.size) {
    if (print.iteration == TRUE) {
      print(i)
    }
    if (cluster.sample == FALSE) {
      resample_ind <- sample(nrow(data), replace = TRUE)
      data_bs <- data[resample_ind, ]
    } else if (cluster.sample == TRUE) {
      data_bs <- cluster.resample(data, cluster.name, size=length(unique(data[,cluster.name])))
    }
    bs_fit.y <- update(ini_fit.y, data = data_bs)
    coef_bs.y <- coef(bs_fit.y)
    if (family.y[1] == "gaussian" & sample.resid.y ==
        FALSE) {
      sd.ref.y <- sd(bs_fit.y$residuals[data_bs[, group] ==
                                          levels(data_bs[, group])[1]])
    }
    if (family.y[1] == "gaussian" & sample.resid.y ==
        TRUE) {
      resid.ref.y <- bs_fit.y$residuals[data_bs[, group] ==
                                          levels(data_bs[, group])[1]]
    }

    if(cluster.mrows==FALSE) {
      bs_fit.m <- update(ini_fit.m, data = data_bs)
      coef_bs.m <- coef(bs_fit.m)
    } else if(cluster.mrows==TRUE) {
      data_bs.m <- do.call(rbind,by(data_bs, list(data_bs[,cluster.name]),FUN=function(x) head(x,1)))
      bs_fit.m <- update(ini_fit.m, data = data_bs.m)
      coef_bs.m <- coef(bs_fit.m)
    }
    if (family.m[1] == "gaussian" & sample.resid.m ==
        FALSE) {
      sd.ref.m <- sd(bs_fit.m$residuals[data_bs[, group] ==
                                          levels(data_bs[, group])[1]])
    }
    if (family.m[1] == "gaussian" & sample.resid.m == TRUE & cluster.mrows == FALSE) {
      resid.ref.m <- bs_fit.m$residuals[data_bs[, group] ==
                                          levels(data_bs[, group])[1]]
    }
    if (family.m[1] == "gaussian" & sample.resid.m == TRUE & cluster.mrows == TRUE) {
      resid.ref.m <- bs_fit.m$residuals[data_bs.m[, group] ==
                                          levels(data_bs.m[, group])[1]]
    }
    data_bs$truegroup <- data_bs[, group]
    for (ii in 1:mc.size) {
      data_mc <- data_bs
      pred_mean_m <- predict(bs_fit.m, data_mc, type = "response")
      n <- length(pred_mean_m)
      if (family.m[1] == "poisson") {
        data_mc[, mediator] <- rpois(n, pred_mean_m)
      }      else if (family.m[1] == "binomial") {
        data_mc[, mediator] <- rbinom(n, 1, pred_mean_m)
      }      else if (family.m[1] == "gaussian" & sample.resid.m ==
                      FALSE) {
        data_mc[, mediator] <- rnorm(n, pred_mean_m,
                                     sd = sd.ref.m)
      }      else if (family.m[1] == "gaussian" & sample.resid.m ==
                      TRUE) {
        data_mc[, mediator] <- pred_mean_m + sample(resid.ref.m,
                                                    n, replace = TRUE)
      }
      temp_nc_m[ii, ] <- tapply(data_mc[, mediator], list(data_mc[,
                                                                  group]), mean, na.rm = TRUE)
      pred_mean_y <- predict(bs_fit.y, data_mc, type = "response")
      data_mc$pred_mean_y <- pred_mean_y
      if (family.y[1] == "poisson") {
        pred_y <- rpois(n, pred_mean_y)
      }      else if (family.y[1] == "binomial") {
        pred_y <- rbinom(n, 1, pred_mean_y)
      }      else if (family.y[1] == "gaussian" & sample.resid.y ==
                      FALSE) {
        pred_y <- rnorm(n, pred_mean_y, sd = sd.ref.y)
      }      else if (family.y[1] == "gaussian" & sample.resid.y ==
                      TRUE) {
        pred_y <- pred_mean_y + sample(resid.ref.y, n,
                                       replace = TRUE)
      }
      # ! append pred_y to data_mc so that the other info in there
      # can also be used by FUN.y
      data_mc$pred_y <- pred_y
      temp_nc_y[ii,] <- FUN.y(data_mc,...)
      data_mc$pred_y <- NA
      data_mc$pred_mean_y <- NA
      data_mc[, group][which(data_mc[, group] %in% levels(data_mc[,
                                                                  group])[-1])] <- levels(data_mc[, group])[1]
      pred_mean_m <- predict(bs_fit.m, data_mc, type = "response")
      if (family.m[1] == "poisson") {
        data_mc[, mediator] <- rpois(n, pred_mean_m)
      }      else if (family.m[1] == "binomial") {
        data_mc[, mediator] <- rbinom(n, 1, pred_mean_m)
      } else if (family.m[1] == "gaussian" & sample.resid.m ==
                 FALSE) {
        data_mc[, mediator] <- rnorm(n, pred_mean_m,
                                     sd = sd.ref.m)
      }       else if (family.m[1] == "gaussian" & sample.resid.m ==
                       TRUE) {
        data_mc[, mediator] <- pred_mean_m + sample(resid.ref.m,
                                                    n, replace = TRUE)
      }
      temp_cf_m[ii, ] <- tapply(data_mc[, mediator], list(data_mc$truegroup),
                                mean, na.rm = TRUE)
      data_mc[, group] <- data_mc$truegroup
      pred_mean_y <- predict(bs_fit.y, data_mc, type = "response")
      data_mc$pred_mean_y <- pred_mean_y
      if (family.y[1] == "poisson") {
        pred_y <- rpois(n, pred_mean_y)
      }      else if (family.y[1] == "binomial") {
        pred_y <- rbinom(n, 1, pred_mean_y)
      }      else if (family.y[1] == "gaussian" & sample.resid.y ==
                      FALSE) {
        pred_y <- rnorm(n, pred_mean_y, sd = sd.ref.y)
      }      else if (family.y[1] == "gaussian" & sample.resid.y ==
                      TRUE) {
        pred_y <- pred_mean_y + sample(resid.ref.y, n,
                                       replace = TRUE)
      }
      # ! append pred_y to data_mc so that the other info in there
      # can also be used by FUN.y
      data_mc$pred_y <- pred_y
      temp_cf_y[ii,] <- FUN.y(data_mc,...)
      data_mc$pred_y <- NA
      data_mc$pred_mean_y <- NA
    }
    if (i == 1) {
      mc_conv_info_m <- apply(temp_nc_m, 2, conv.mean)
      mc_conv_info_y <- apply(temp_nc_y, 2, conv.mean)
    }
    out_nc_m[i, ] <- apply(temp_nc_m, 2, mean, na.rm = TRUE)
    out_cf_m[i, ] <- apply(temp_cf_m, 2, mean, na.rm = TRUE)
    out_nc_y[i, ] <- apply(temp_nc_y, 2, FUN=mean,na.rm=TRUE)
    out_cf_y[i, ] <- apply(temp_cf_y, 2, FUN=mean,na.rm=TRUE)
  }
  return(list(out_nc_m = out_nc_m,
              out_cf_m = out_cf_m,
              out_nc_quantile_m = apply(out_nc_m, 2, quantile, c(alpha/2, 0.5, 1 - alpha/2),na.rm=TRUE),
              out_cf_quantile_m = apply(out_cf_m, 2, quantile, c(alpha/2, 0.5, 1 - alpha/2),na.rm=TRUE),
              out_nc_y = out_nc_y,
              out_cf_y = out_cf_y,
              out_nc_quantile_y = apply(out_nc_y, 2, quantile, c(alpha/2, 0.5, 1 - alpha/2),na.rm=TRUE),
              out_cf_quantile_y = apply(out_cf_y, 2, quantile, c(alpha/2, 0.5, 1 - alpha/2),na.rm=TRUE),
              mc_conv_info_m = mc_conv_info_m,
              mc_conv_info_y = mc_conv_info_y))
}
